naming/dns_resolver.go

```diff
@@ -0,0 +1,312 @@
+			if err != nil {
+				return "", "", err
+			}
+		}
+	}
+	// Keep consistent with net.Dial(): If the host is empty, as in ":80", the local system is assumed.
+	if host == "" {
+		host = "localhost"
+	}
+	return host, port, nil
+}
+
+// Resolve creates a watcher that watches the name resolution of the target.
+func (r *dnsResolver) Resolve(target string) (Watcher, error) {
+	host, port, err := setHostPort(target)
+	if err != nil {
+		return nil, err
+	}
+
+	if net.ParseIP(host) != nil {
+		ipWatcher := &IPWatcher{
+			r:      r,
+			target: target,
+			host:   host,
+			port:   port,
+			IPAddr: make(chan *Update, 1),
+			done:   make(chan struct{}),
+		}
+		host, _ = formatIP(host)
+		ipWatcher.IPAddr <- &Update{Op: Add, Addr: host + ":" + port}
+		return ipWatcher, nil
+	}
+
+	return &dnsWatcher{
+		r:      r,
+		target: target,
+		host:   host,
+		port:   port,
+		done:   make(chan struct{}),
+	}, nil
+}
+
+// dnsWatcher watches for the name resolution update for a specific target
+type dnsWatcher struct {
+	r      *dnsResolver
+	target string
+	host   string
+	port   string
+	// The latest resolved address list
+	curAddrs []*Update
+	// done channel is closed to notify Next() to exit.
+	done chan struct{}
+}
+
+// IPWatcher watches for the name resolution update for an IP address.
+type IPWatcher struct {
+	r      *dnsResolver
+	target string
+	host   string
+	port   string
+	IPAddr chan *Update
+	// done channel is closed to notify Next() to exit.
+	done chan struct{}
+}
+
+// Next returns the adrress resolution Update for the target. For IP address,
+// the resolution is itself, thus polling name server is unncessary. Therefore,
+// Next() will return an Update the first time it is called, and will be blocked
+// for all following calls as no Update exisits until watcher is closed.
+func (i *IPWatcher) Next() ([]*Update, error) {
+	select {
+	case u := <-i.IPAddr:
+		return []*Update{u}, nil
+	case <-i.done:
+		return nil, errors.New(watcherClose)
+	}
+}
+
+// Close closes the IPWatcher.
+func (i *IPWatcher) Close() {
+	close(i.done)
+}
+
+// AddressType indicates the address type returned by name resolution.
+type AddressType uint8
+
+const (
+	// Backend indicates the server is a backend server.
+	Backend AddressType = iota
+	// GRPCLB indicates the server is a grpclb load balancer.
+	GRPCLB
+)
+
+// AddrMetadataGRPCLB contains the information the name resolver for grpclb should provide. The
+// name resolver used by the grpclb balancer is required to provide this type of metadata in
+// its address updates.
+type AddrMetadataGRPCLB struct {
+	// AddrType is the type of server (grpc load balancer or backend).
+	AddrType AddressType
+	// ServerName is the name of the grpc load balancer. Used for authentication.
+	ServerName string
+}
+
+// compileUpdate compares the old resolved addresses and newly resolved addresses,
+// and generates an update list
+func compileUpdate(oldAddrs []*Update, newAddrs []*Update) []*Update {
```

Comment:
Could this be simplified by using a map instead of walking through the slices?  I.e.:

```go
update := map[string]*Update
for _, a := range newAddrs {
  update[a.Addr] = &Update{Addr: a.Addr, Op: Add}
}
for _, a := range oldAddrs {
  if _, ok := update[a.Addr]; ok {
    delete(update[a.Addr])
    continue
  }
  update[a.Addr] = &Update{Addr: a.Addr, Op: Delete}
}
// convert update into a slice and return it
```

You can probably simplify further by making `oldAddrs` and `newAddrs` sets (`map[string]bool`) instead of a slice of `Update`s.

In either case, you also wouldn't need to sort the address lists, either.